lib/streamlit/runtime/app_session.py

```diff
@@ -275,6 +280,9 @@ def _enqueue_forward_msg(self, msg: ForwardMsg) -> None:
         if self._debug_last_backmsg_id:
             msg.debug_last_backmsg_id = self._debug_last_backmsg_id
 
+        msg.metadata.sequence_number = self._forward_msg_counter
+        self._forward_msg_counter += 1
```

Comment:
Could this part be thread-unsafe, if user invokes threads from the streamlit app that sends ForwardMsg-s (Is it true in the case of st.fragment?)

Adding Lock could be a solution
Although another idea could be to move it to self._browser_queue ForwardMsgQueue() / runtime _send_message to have a sequential place where we handle this

Reply:
There's no possible fragment issue here since fragments don't run concurrently + are run by our normal script runner. There may, however, potentially be an issue if a `ForwardMsg` is enqueued by both the user script and main script (say, in response to a `BackMsg` requesting `FileUploadUrls`).

I don't think adding a lock would be a great solution given the added overhead around grabbing/releasing a lock before every `ForwardMsg` sent (even if we don't really expect any lock contention), but I think an alternative would be to do get/increment our counter in a place where we're sure only a single thread can ever be running. That is, we can get/increment the counter right when sending messages to the browser as only the main eventloop thread is able to do that.

Reply:
Yeah, agree

That place could be Runtime `_send_message`, in fact, we already increment there a `session_info.script_run_count` (needed for removing old ForwardMsgCache entries)

https://github.com/streamlit/streamlit/blob/c73e2262d1836196c6ddd18eb9978e8f9238617f/lib/streamlit/runtime/runtime.py#L737